### PR TITLE
Reset first-iteration stats when logging every iteration

### PR DESCRIPTION
**Pull request by Sharllysichen**
## Description

Reset logging accumulators after the first iteration when `log_interval == 1`.

The first iteration is logged separately for initialization statistics and is normally kept in the accumulator so that the first regular logging window contains a complete `log_interval` iterations. However, when `log_interval` is 1, this causes the second iteration's logged loss and timing statistics to include both the first and second iterations.

This change preserves the existing behavior for `log_interval > 1` while ensuring that every log represents a single iteration when `log_interval == 1`.

## Testing

- Verified that `log_interval == 1` resets statistics after the first iteration.
- Verified that the existing behavior is unchanged for `log_interval > 1`.

### CODE REVIEW

**Review by sylvesterkaczmarek — COMMENTED** [review]
I checked this against the accumulator/reset path in `training_log()`. The first iteration is logged regardless of the normal interval (`iteration % args.log_interval == 0 or is_first_iteration`), while `should_reset` controls the interval timer reset, per-loss accumulator reset, advanced/skipped/NaN counters, and the stdout timer reset.

For `log_interval == 1`, keeping `should_reset=False` on the first iteration therefore leaves that iteration in all of those accumulators, so the second log covers two iterations. Making the first log reset only in the interval-1 case isolates iteration 2 as intended. For every larger interval, the existing behavior is preserved and iteration 1 remains part of the first regular logging window.

I don't see a blocking issue with this change.

**Review by wujingyue — COMMENTED** [review]
I'm not sure I understand the problem. Do you have a repro?

**Comment by Sharllysichen** [comment]
@wujingyue Thanks for asking. I reproduced the issue with three runs using the same
training configuration and random seed. The only differences between the runs
were `log_interval` and the proposed `should_reset` condition.

With the current code:

```python
should_reset = not is_first_iteration
```
and log_interval=1, the relevant output is:
```
iteration 1 | elapsed time per iteration (ms): 69910.0 | lm loss: 1.104833E+01
iteration 2 | elapsed time per iteration (ms): 35351.9 | lm loss: 1.137185E+01
iteration 3 | elapsed time per iteration (ms):   780.5 | lm loss: 1.167506E+01
```
The iteration-2 loss is not the loss from iteration 2 alone. The patched run shows that the iteration-2 loss is 1.169537E+01, while the current code reports exactly the average of iterations 1 and 2:
```
(1.104833E+01 + 1.169537E+01) / 2
= 1.137185E+01
```
This occurs because the first iteration is logged with is_first_iteration=True, which makes should_reset=False. Therefore,
total_loss_dict and the interval timer are not reset after the first log. On iteration 2, advanced iterations becomes 2 and the logged loss is calculated from both iterations:
```
(total_loss_iter_1 + total_loss_iter_2) / 2
```
The timing metric is affected in the same way. The iteration-2 value of 35351.9 ms still includes the first iteration's initialization overhead. From iteration 3 onward, the statistics become per-iteration again because iteration 2 finally resets the accumulators. 

As a control, with the current code and log_interval=2, I get:
```
iteration 1 | elapsed time per iteration (ms): 34904.8 | lm loss: 1.104833E+01
iteration 2 | elapsed time per iteration (ms): 17863.8 | lm loss: 1.137185E+01
iteration 4 | elapsed time per iteration (ms):   782.0 | lm loss: 1.168838E+01
```
The iteration-2 loss is again 1.137185E+01. This confirms that the log_interval=1 run currently treats its second log like the first two-iteration window of a log_interval=2 run.

For log_interval=2, retaining the first iteration is expected because the first regular logging window contains iterations 1 and 2. The proposed change preserves that behavior.

 With the proposed code:
```python
should_reset = not is_first_iteration or args.log_interval == 1
```
and log_interval=1, the relevant output becomes:
```
iteration 1 | elapsed time per iteration (ms): 32185.2 | lm loss: 1.104833E+01
iteration 2 | elapsed time per iteration (ms):   805.6 | lm loss: 1.169537E+01
iteration 3 | elapsed time per iteration (ms):   766.7 | lm loss: 1.167503E+01
```
Iteration 2 now reports its own loss (1.169537E+01) and iteration time (805.6 ms) instead of averages that include iteration 1 and its initialization overhead.

Therefore, this change is limited to the log_interval=1 edge case. For log_interval>1, the condition still evaluates to false on the first iteration, preserving the existing first-window behavior.

**Review by wujingyue — DISMISSED** [review]
(Deleted)

File: megatron/training/training.py [hunk 1]
```diff
@@ -3335,8 +3335,9 @@ def training_log(
     one_logger = get_one_logger()
     energy_monitor = get_energy_monitor()
 
-    # On first iteration, log stats but don't reset accumulators so normal interval stats remain accurate.
-    should_reset = not is_first_iteration
+    # On first iteration, log stats but don't reset accumulators so normal interval stats remain accurate. When logging every iteration, reset immediately so the
+    # second iteration does not include statistics from the first iteration.
+    should_reset = not is_first_iteration or args.log_interval == 1
```
Comment (wujingyue) [inline megatron/training/training.py]: I'd recommend something like
```
  if args.log_interval == 1:
      should_reset = True
  else:
      # Keep the startup iteration in the first regular logging window.
      should_reset = not is_first_iteration
```

It's equivalent but more readable. Humans are pretty bad at composite conditions.

**Review by wujingyue — COMMENTED** [review]
LGTM overall

Also, is it possible to add a unit test? This condition is tricky to get right. We should probably have some coverage so people don't accidentally change it back or wrong in a different way.

File: megatron/training/training.py [hunk 2]
```diff
@@ -3309,6 +3309,15 @@ def _save_state_dict(attr_name, label):
...
 
+def _should_reset_logging_interval(is_first_iteration: bool, log_interval: int) -> bool:
+    """Return whether logging accumulators should be reset after this iteration."""
+    if log_interval == 1:
+        # Every iteration is a complete logging window, including the startup iteration.
+        should_reset = True
+    else:
+        # Keep the startup iteration in the first regular logging window.
+        should_reset = not is_first_iteration
+    return should_reset
```
Comment (wujingyue) [inline megatron/training/training.py]:
```suggestion
    if log_interval == 1:
        # Every iteration is a complete logging window, including the startup iteration.
        return True

    # Keep the startup iteration in the first regular logging window.
    return not is_first_iteration
```

File: tests/unit_tests/training/test_training_log.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        (False, 2, True),
+    ],
+    ids=[
+        "first-iteration-single-iteration-window",
+        "first-iteration-multi-iteration-window",
+        "later-iteration-single-iteration-window",
+        "later-iteration-multi-iteration-window",
+    ],
+)
+def test_should_reset_logging_interval(is_first_iteration, log_interval, expected):
```
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py]: https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html

Sorry, I wasn't clear about what test I was expecting. 

Try to focus on behavior. For example, run a real training loop with log_interval==1 and capture the output losses from stdout and check those match your expectation.

**Review by wujingyue — CHANGES_REQUESTED** [review]

**Comment by wujingyue** [comment]
Also, please fix DCO

File: megatron/training/training.py (hunk 1 above)
Comment (Sharllysichen) [reply megatron/training/training.py]: Thanks for the suggestion. I updated the implementation following your recommendation:

```python
# Determine whether the current iteration completes the logging window.
if args.log_interval == 1:
    # Every iteration is a complete logging window, including the startup iteration.
    should_reset = True
else:
    # Keep the startup iteration in the first regular logging window.
    should_reset = not is_first_iteration
```
I also added a parameterized unit test that exercises training_log() directly and verifies the timer reset behavior for all four relevant cases:
- First iteration with log_interval == 1: reset
- Later iteration with log_interval == 1: reset
- First iteration with log_interval > 1: do not reset
- Later iteration with log_interval > 1: reset

The focused test passes:
```text
4 passed
```
I also ran tools/autoformat.sh; Black, isort, pylint, Ruff, and mypy all pass.

File: tests/unit_tests/training/test_training_log.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+from unittest.mock import MagicMock
+
+import pytest
+
+from megatron.training.training import training_log
+
+
+def _make_args(log_interval: int):
+    return SimpleNamespace(
+        log_interval=log_interval,
```
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py]: Why does this class contain so many attributes unrelated to `log_interval`?

File: tests/unit_tests/training/test_training_log.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+):
+    args = _make_args(log_interval)
+
+    timer = MagicMock()
+    timer.elapsed.return_value = 1.0
+
+    timers = MagicMock()
+    timers.return_value = timer
+
+    monkeypatch.setattr("megatron.training.training.get_args", lambda: args)
```
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py]: Why is the test so complicated for just collecting losses?

**Review by wujingyue — CHANGES_REQUESTED** [review]

File: tests/unit_tests/training/test_training_log.py (hunk 5 above)
Comment (Sharllysichen) [reply tests/unit_tests/training/test_training_log.py]: Thank you for the feedback. I agree that the previous test setup made the behavior harder to follow. I’ve revised it to address your comments.

The extra attributes were needed by the code paths accessed by training_log(), but manually maintaining them in a SimpleNamespace made the setup unnecessarily verbose. The updated test uses Megatron’s argument parser and initialization helpers with a minimal test configuration.

I also removed the timer mocks and assertions about internal reset operations. The test now runs a small, deterministic PyTorch SGD loop, calls the actual training_log() function at each step, and checks the iteration numbers and loss values captured from stdout. This makes the expected logging behavior explicit.

The six parameterized cases cover log_interval values of 1, 2, and 3, with iteration numbering starting from both zero and a nonzero offset. They check that each log contains the current iteration’s loss when log_interval == 1, while the first iteration remains included in the first regular logging window for larger intervals.

To verify the regression coverage, I tested both versions locally:
- With the fix: 6 passed.
- With the condition temporarily reverted to should_reset = not is_first_iteration: 2 failed, 4 passed. Both log_interval == 1 cases failed, while the interval-2 and interval-3 cases still passed.

The fix has been restored, and I’ve also corrected the missing commit sign-offs. DCO is now passing.

File: tests/unit_tests/training/test_training_log.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        validate_args(args)
+
+        # FLOPs reporting needs this; the scalar model needs no tokenizer.
+        args.padded_vocab_size = 128
+        set_global_variables(args, build_tokenizer=False)
+        yield args
+    finally:
+        destroy_global_vars()
+        destroy_num_microbatches_calculator()
+        Utils.destroy_model_parallel()
```
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py:49]: `def teardown` to reduce indentation.

File: tests/unit_tests/training/test_training_log.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        args.world_size = torch.distributed.get_world_size()
+        validate_args(args)
+
+        # FLOPs reporting needs this; the scalar model needs no tokenizer.
+        args.padded_vocab_size = 128
+        set_global_variables(args, build_tokenizer=False)
+        yield args
+    finally:
+        destroy_global_vars()
+        destroy_num_microbatches_calculator()
```
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py:48]: Why?

File: tests/unit_tests/training/test_training_log.py
Comment (wujingyue) [inline tests/unit_tests/training/test_training_log.py:1]: cc @maanug-nv 

Do you know the minimal way to test training log? The test as is seems quite complicated. I would be surprised if we can't write an easier unit-test to train a toy model end-to-end and examine the losses via `training_log`. 

Given the test as is, I may prefer removing it entirely because of https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html

Comment (Sharllysichen) [reply tests/unit_tests/training/test_training_log.py:1]: @wujingyue Thank you for clarifying. I understand your concern: although the assertions check the logged losses, the test still requires substantial setup and manually drives training_log(), which makes it harder to maintain.

I haven’t yet found a suitable existing toy-model test or helper that exercises this through the normal Megatron training flow. Given the complexity of the current test, I’d suggest removing it and keeping this PR focused on the fix, with the reproduction and validation results documented in the discussion. Would you be comfortable proceeding without adding a unit test in this PR?
